File: lib/underwritingGraph.js
async function runUnderwritingGraph({ messages, model, apiKey, guideInstructions, callOpenAI }) {
  ensureWebStreamGlobals();

  const { Annotation, END, START, StateGraph } = await import("@langchain/langgraph");

  const GraphState = Annotation.Root({
    messages: Annotation({
      reducer: (_left, right) => right,
      default: () => []
    }),
    model: Annotation({
      reducer: (_left, right) => right,
      default: () => model
    }),
    apiKey: Annotation({
      reducer: (_left, right) => right,
      default: () => apiKey
    }),
    guideInstructions: Annotation({
      reducer: (_left, right) => right,
      default: () => []
    }),
    reply: Annotation({
      reducer: (_left, right) => right,
      default: () => ""
    }),
    responseId: Annotation({
      reducer: (_left, right) => right,
      default: () => null
    })
  });

  async function callUnderwritingModel(state) {
    const response = await callOpenAI({
      apiKey: state.apiKey,
      guideInstructions: state.guideInstructions,
      model: state.model,
      messages: state.messages
    });

    return {
      reply: response.reply,
      responseId: response.id
    };
  }

  const graph = new StateGraph(GraphState)
    .addNode("underwriting_model", callUnderwritingModel)
    .addEdge(START, "underwriting_model")
    .addEdge("underwriting_model", END)
    .compile();

  return graph.invoke({
    messages,
    model,
    apiKey,
    guideInstructions
  });
}

function ensureWebStreamGlobals() {
  if (
    typeof AbortSignal !== "undefined" &&
    AbortSignal.prototype &&
    typeof AbortSignal.prototype.throwIfAborted !== "function"
  ) {
    AbortSignal.prototype.throwIfAborted = function throwIfAborted() {
      if (this.aborted) {
        throw this.reason || new Error("The operation was aborted.");
      }
    };
  }

  if (
    typeof globalThis.ReadableStream !== "undefined" &&
    typeof globalThis.WritableStream !== "undefined" &&
    typeof globalThis.TransformStream !== "undefined"
  ) {
    return;
  }

  const {
    ReadableStream,
    TransformStream,
    WritableStream
  } = require("stream/web");

  globalThis.ReadableStream = globalThis.ReadableStream || ReadableStream;
  globalThis.TransformStream = globalThis.TransformStream || TransformStream;
  globalThis.WritableStream = globalThis.WritableStream || WritableStream;
}

module.exports = {
  runUnderwritingGraph
};

File: backend/agents/underwriting_graph.py
from typing import Any, Callable, TypedDict


class UnderwritingState(TypedDict, total=False):
    messages: list[dict[str, str]]
    model: str
    api_key: str
    guide_instructions: list[str]
    underwriter_notes: list[str]
    reply: str
    response_id: str | None


def run_underwriting_graph(
    messages: list[dict[str, str]],
    model: str,
    api_key: str,
    guide_instructions: list[str],
    underwriter_notes: list[str],
    call_openai: Callable[..., dict[str, Any]],
):
    try:
        from backend.agents.openai_agents_sdk import run_openai_agents_sdk

        return run_openai_agents_sdk(
            messages=messages,
            model=model,
            api_key=api_key,
            guide_instructions=guide_instructions,
            underwriter_notes=underwriter_notes,
        )
    except ModuleNotFoundError:
        pass

    try:
        return run_langgraph(
            messages,
            model,
            api_key,
            guide_instructions,
            underwriter_notes,
            call_openai,
        )
    except ModuleNotFoundError:
        return call_underwriting_model(
            {
                "messages": messages,
                "model": model,
                "api_key": api_key,
                "guide_instructions": guide_instructions,
                "underwriter_notes": underwriter_notes,
            },
            call_openai,
        )


def run_langgraph(messages, model, api_key, guide_instructions, underwriter_notes, call_openai):
    from langgraph.graph import END, START, StateGraph

    def underwriting_model_node(state: UnderwritingState):
        return call_underwriting_model(state, call_openai)

    graph = StateGraph(UnderwritingState)
    graph.add_node("underwriting_model", underwriting_model_node)
    graph.add_edge(START, "underwriting_model")
    graph.add_edge("underwriting_model", END)
    compiled = graph.compile()

    result = compiled.invoke(
        {
            "messages": messages,
            "model": model,
            "api_key": api_key,
            "guide_instructions": guide_instructions,
            "underwriter_notes": underwriter_notes,
        }
    )
    result["framework"] = "python-langgraph"
    return result


def call_underwriting_model(state, call_openai):
    response = call_openai(
        api_key=state["api_key"],
        model=state["model"],
        messages=state["messages"],
        guide_instructions=state.get("guide_instructions", []),
        underwriter_notes=state.get("underwriter_notes", []),
    )
    return {
        **state,
        "reply": response.get("reply", ""),
        "response_id": response.get("id"),
        "framework": "python-graph-fallback",
    }
